### ISSUE TITLE
Performance: Contour suffers 2-3x slowdown programming Envoys under heavy churn

### ISSUE THREAD

**Issue by mattmoor** [projectcontour/contour#2857]
**What steps did you take and what happened:**

tl;dr Run Knative's "scale-100" test.

I've been using Knative's Prow to run this for me.  
1. `ko publish ./cmd/contour`
2. Replace references to `docker.io/projectcontour/contour:*` with digest in [here](https://github.com/knative/serving/blob/master/third_party/contour-latest/contour.yaml)
3. Open a PR, this will automatically run `pull-knative-serving-contour-latest`

To access logs...
1. Click the "Details" on the `pull-knative-serving-contour-latest` check.
2. At the top of the next page click on "Artifacts".
3. Download `build-log.txt`

I generally scan for `scale-100` and snip everything above that (`scale-10` typically runs first), then you can `grep time-to-200 {build-log.txt} | sort`

**What did you expect to happen:**

See the timings below, but I'd love to see Contour somewhere closer in latency to where it is with status disabled 🤩 

**Anything else you would like to add:**

With 1.4 I see (just the tail latencies of 100 ksvcs):

```
    scale_test.go:38: "time-to-200" took 2m13.922577766s
    scale_test.go:38: "time-to-200" took 2m15.546537895s
    scale_test.go:38: "time-to-200" took 2m17.604262618s
    scale_test.go:38: "time-to-200" took 2m2.548284264s
    scale_test.go:38: "time-to-200" took 2m24.920357586s
    scale_test.go:38: "time-to-200" took 2m35.001542654s
    scale_test.go:38: "time-to-200" took 2m4.284511208s
    scale_test.go:38: "time-to-200" took 2m5.082870703s
    scale_test.go:38: "time-to-200" took 2m6.411782863s
```

With Contour 1.5+ these climb **over 5 minutes**.

With Contour 1.8 and status updates manually disabled (these lines commented):

```
    scale_test.go:38: "time-to-200" took 45.933159712s
    scale_test.go:38: "time-to-200" took 46.045489619s
    scale_test.go:38: "time-to-200" took 48.399353288s
    scale_test.go:38: "time-to-200" took 48.469338024s
    scale_test.go:38: "time-to-200" took 48.661313485s
    scale_test.go:38: "time-to-200" took 48.662251973s
    scale_test.go:38: "time-to-200" took 50.3964567s
    scale_test.go:38: "time-to-200" took 52.196902407s
    scale_test.go:38: "time-to-200" took 52.359932905s
```

With [this](https://github.com/projectcontour/contour/pull/2854) patch to 1.8:

```
    Scale_test.go:38: "time-to-200" took 2m3.506789091s
    scale_test.go:38: "time-to-200" took 2m31.205368141s
    scale_test.go:38: "time-to-200" took 2m32.195546552s
    scale_test.go:38: "time-to-200" took 2m34.998376039s
    scale_test.go:38: "time-to-200" took 2m53.295739778s
    scale_test.go:38: "time-to-200" took 2m55.178364304s
    scale_test.go:38: "time-to-200" took 2m56.109412892s
    scale_test.go:38: "time-to-200" took 3m11.610176956s
```

**Environment:**

- Contour version: (see details above)
- Kubernetes version:  GKE 1.16

**Comment by stevesloka** [comment]
@mattmoor do you bump the version of Envoy along with the upgrade (i.e. using Envoy v1.15 vs something older?). I don't think it's going to give you that large of a bump, was just more-so curious.

**Comment by mattmoor** [comment]
We basically grab your release configs and apply some patches, so we use whatever Envoy you folks reference.

**Comment by mattmoor** [comment]
So for the patched case, above I had a bug (missing RBAC) with it fixed, things are worst (a couple failures).

My statement about the timeout being 5 minutes was also wrong, it's 5 minutes total, and we throttle the creation rate to 50 at a time, so I think it's better to think of the ceiling as 2.5 minutes.

Looking at one case that failed with the fixed path, it took about 60s for the httpproxy we use for endpoint probe to become ready followed by over two minutes for the actual programming to become ready after that.

**Comment by mattmoor** [comment]
As another interesting datapoint, if I take the patch above and add in an (unreasonably high) rate limit increase the latencies reduce to:

```
    scale_test.go:38: "time-to-200" took 1m20.390800429s
    scale_test.go:38: "time-to-200" took 1m22.186353402s
    scale_test.go:38: "time-to-200" took 1m30.472498018s
    scale_test.go:38: "time-to-200" took 1m32.265027868s
    scale_test.go:38: "time-to-200" took 1m33.103806322s
    scale_test.go:38: "time-to-200" took 1m33.988958365s
    scale_test.go:38: "time-to-200" took 1m34.670913009s
    scale_test.go:38: "time-to-200" took 1m35.110076242s
    scale_test.go:38: "time-to-200" took 1m37.484821359s
    scale_test.go:38: "time-to-200" took 1m45.195982691s
    scale_test.go:38: "time-to-200" took 1m48.175498773s
```

So we know:
1. Eliminating the status update massively improves latency (see original comment)
2. Increasing our client-side quota (to be unreasonably high) improves latency (directly above)

This makes me think that we're doing a fairly high volume of patches, so I am going to try instrumenting the code a bit more to get a better understanding of what's going on here.

**Comment by mattmoor** [comment]
So chatted with @youngnick a bit in slack, and here's a theory about why client rate-limiting might be affecting the latency of Envoy programming.

When Contour rebuilds the DAG, the leader sets the status on http proxies before returning:
https://github.com/projectcontour/contour/blob/2e80af0f34bf674ac488e0d05ce4c3128c09bd23/internal/contour/handler.go#L218

For each http proxy, this calls SetStatus:
https://github.com/projectcontour/contour/blob/2e80af0f34bf674ac488e0d05ce4c3128c09bd23/internal/contour/handler.go#L229

SetStatus synchronously calls Update:
https://github.com/projectcontour/contour/blob/2e80af0f34bf674ac488e0d05ce4c3128c09bd23/internal/k8s/status.go#L114

Update sends the status update on the channel:
https://github.com/projectcontour/contour/blob/2e80af0f34bf674ac488e0d05ce4c3128c09bd23/internal/k8s/statusupdater.go#L230

**This channel is unbuffered, so that line will block until another thread is ready to receive off of that channel.**

So if this loop is blocked on the previous update (e.g. due to client-side rate limiting) then we will see potentially long pauses between receives here:
https://github.com/projectcontour/contour/blob/2e80af0f34bf674ac488e0d05ce4c3128c09bd23/internal/k8s/statusupdater.go#L74

To put this in perspective, if we have 100 status updates to perform, and the client is rate limiting us at 5 QPS (the default) then we will see a 20s pause at the tail end of the DAG rebuild before it completes.

What this should probably be doing is `Update` should "enqueue" the status update into a [workqueue](https://godoc.org/k8s.io/client-go/util/workqueue) and the loop in `Start` should drain from the workqueue instead.

This should make the status portion of the DAG rebuild take effective no time to hand-off, and will probably achieve nearly the theoretical best timings above 🤩

**Comment by youngnick** [comment]
This issue was fixed in [release 1.8.1](https://github.com/projectcontour/contour/releases/tag/v1.8.1). Thanks @mattmoor for all your work in getting to the bottom of it!
